Refactor querying Distgit/Pagure owners (and caching)

nphilipp opened the pull request:
Apart from reworking how FMN accesses Pagure/Distgit, plus quite some minor improvements and fixes. Sorry for the huge drop! 😂

nphilipp committed:
Make async client caching work between processes

With cashews, the string value of an object is used to compute the key
with which a cached method result is looked up. The default
implementation of __str__() includes the id of the object, which
prevents different processes to access the same cached objects.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Add configuration and helper for cache arguments

We want to make the TTL of cached items configurable, this change
implements helper factories to produce callables which return the
respective configured values for use in the cashews.cache() decorator.

In the course, adapt some pieces in the existing cache configuration to
enable this.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Fix dumping/loading of the configuration structure

Previously, validating the Settings model computed some compound fields,
but they were unknown to the model. Therefore dumping the structure to a
dict and recreating it would fail.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Use relative imports consistently

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Add more type hints

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Fix fractional hour values for last messages

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Guard against recursive pagination

If the result of determine_next_page_params() repeats this throws an
exception, assuming that the backend would send the client in a loop
requesting the same page over and over.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Handle Pagure pagination query parameters better

The `next_url` parameter returned for paginated responses contains query
parameters. In order to avoid having the same keys with conflicting
values (such as `page`) in the supplied `params` dictionary, extract
them from `next_url` and store them in `params`.

Signed-off-by: Nils Philippsen <nils@redhat.com>

abompard commented on fmn/api/handlers/misc.py:
```diff
@@ -80,8 +86,7 @@ async def get_artifacts(
     namespaces = [at.value for at in ArtifactType]
-    projects = await distgit_client.get_projects(pattern=name)
-    for project in projects:
+    for project in await distgit_client.get_projects(pattern=f"*{name}*"):
```
Are you sure we need to do a substring search by padding asterisks? It feels like we may end up with more projects than we expect.

nphilipp replied:
This was done in the (`fmn.rules`) distgit client previously, I moved that out of the implementation in `fmn.backends`. I'm fine with removing it altogether or move this detail into the frontend (people will probably want to search by substring, I know I want to).

abompard commented on fmn/rules/services/distgit.py:
```diff
@@ -19,16 +19,17 @@ def __init__(self, url):
+    def get_user_owners(self, artifact_type: str, artifact_name: str) -> list[str]:
+        return self.proxy.get_project_users(
+            f"{artifact_type}/{artifact_name}",
+            PagureRole.OWNER | PagureRole.ADMIN | PagureRole.COMMIT,
```
Huh, I did not, know that OR'ed IntEnums could be iterated on to get the components!

In fact, after some research, it seems to be a new feature of Python 3.11. Can you confirm? I think we want to target 3.10, no?

nphilipp replied:
You can iterate over the type and get the constituent enum values, this is what's done in `PagureAsyncProxy.get_project_users()` and it seems to work in Python 3.10, too (same output in Python 3.11):

```
nils@makake:~> python3.10
Python 3.10.9 (main, Dec 20 2022, 00:00:00) [GCC 12.2.1 20221121 (Red Hat 12.2.1-4)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from enum import IntEnum, auto
>>> class Foo(IntEnum):
...     bar = auto()
...     baz = auto()
...     gna = auto()
... 
>>> Foo.bar | Foo.baz 
3
>>> for x in Foo:
...     print(x)
... 
Foo.bar
Foo.baz
Foo.gna
>>>
```

nphilipp committed:
Enforce relative imports in CI

Also, clarify why the Alembic env.py module has to use absolute imports.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Backends: Distinguish base URL and API namespace

Additionally, add (cached) base_url_with_trailing_slash property which
can be used to tell apart messages related to different instances of the
same type of service (e.g. dist-git Pagure vs. pagure.io).

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Refactor querying DistGit/Pagure

This consolidates code accessing DistGit/Pagure in PagureAsyncProxy
which was previously strewn across the API and consumer.

In the course, cache the results of the respective proxy methods, so
all consumers of the proxy can benefit.

Fixes: #689

Signed-off-by: Nils Philippsen <nils@redhat.com>

abompard commented on .github/workflows/main.yaml:
```diff
@@ -1,6 +1,27 @@
+      - name: Enforce relative imports in package
+        run: >
+             find fmn -name \*.py -a \! -path fmn/database/migrations/env.py -print0
+             | xargs -0 absolufy-imports --never
```
Why is the Alembic script excluded and then handled differently? I haven't (easily) found what the `--never` switch does.

nphilipp replied:
- Alembic imports the `env.py` file and migration scripts in a way that they don’t know they are inside the `fmn` package, any relative imports outside of the `fmn/database/migrations` directory would fail. In `env.py` there’s a comment about it, I’ll add one here, too.
- The default mode of `absolufy-imports` is to enfore imports are absolute instead of relative, `--never` does the opposite.

abompard commented on fmn/backends/pagure.py:
```diff
@@ -30,5 +55,187 @@ def determine_next_page_params(self, url: str, params: dict, result: dict) -> Ne
+            async for project in self.get_paginated(
+                f"/group/{name}", params=params, payload_field="projects"
+            ):
+                if (fullname := project["fullname"]) in seen_fullnames:
```
Ha, I think that's the first time I see `:=` used! :-)

nphilipp replied:
Coming from C and the like, I was missing it so much in Python! 🙂

abompard commented on fmn/backends/pagure.py:
```diff
@@ -30,5 +55,187 @@ def determine_next_page_params(self, url: str, params: dict, result: dict) -> Ne
+                async for key, _ in cache.get_match(pattern)
+            ]
+        else:  # usergroup == "group"
+            # Messages about changes to project users
```
to project *groups*, no?

nphilipp replied:
Oops, of course!

abompard reviewed (commented):
That's a lot of work! Thanks!
Only minor comments. There are some parts with invalidation and the computation of key templates that I'm not sure I understood entirely, so I'll read it over again, but I'm happy testing it on staging anyway.

abompard reviewed (approved):
Alright, let's merge this and see what happens :-D
